### PR TITLE
More csrf

### PR DESCRIPTION
**Pull request by sravfeyn**
Best viewed commit by commit. First 3 commits are added by looking at csrf-failures from logs.

For the 4th commit, I did a find for all post HTML forms and tried to add `csrf_token`. I have tested at least around 70% of these. I have added csrf to HTML post forms those that I was confident have the right template-context (so that rendered is aware of csrf_token value) and that target internal URLs. There are still some forms to go, but would like to get these in and see if #of csrf failure go down for complete release of csrf.

@esoergel

### CODE REVIEW

**Comment by sravfeyn** [comment]
cc @czue

**Comment by czue** [comment]
:+1: will let @esoergel review closely.

any issues you foresee that could happen with adding it to an untested form?

**Comment by sravfeyn** [comment]
The only issue that I see is that if a form is rendered using `render_to_string` (or some of its cousins) without taking a RequestContext (which knows csrf_token) explicitly as one of its arguments, the page will not have a csrf_token (hence raising csrf failure) or will result in a TemplateError. The [second commit](https://github.com/dimagi/commcare-hq/commit/913885205c422d2458bafeea9f1ad918a46571ed) is an example of fixing such case. Though such cases are very few (I found one such case till now), so I am okay letting such few cases 500 and actively fix them, instead of waiting for csrf-fail logs and then fixing them.

**Comment by czue** [comment]
(seems reasonable to me)

File: corehq/apps/hqwebapp/utils.py [hunk 1]
```diff
@@ -198,3 +198,15 @@ def submenu_dropdown_dict(title, url, menu):
...
+def csrf_inline(request):
+    """
+    Returns "<input type='hidden' name='csrfmiddlewaretoken' value='<csrf-token-value>' />",
+    same as csrf_token template tag, but a shortcut without needing a Template or Context explicitly.
+
+    Useful for adding inline forms in messages for e.g. while showing an "'undo' Archive Form" message
+    """
+    from django.template import Template, RequestContext
+    node = "{% csrf_token %}"
+    return Template(node).render(RequestContext(request))
```
Comment (esoergel) [inline corehq/apps/hqwebapp/utils.py]: Very nice!

**Comment by esoergel** [comment]
Looks good to me.  Have you looked at crispy forms?  Do those include csrf tags by default?

**Comment by sravfeyn** [comment]
Yeah, cripsy_forms by default include csrf. There is a setting to make it not so, if we want per case basis

**Pull request merged**